### ISSUE TITLE
Compiler error when calling operator() with no arguments in Buffer<void>

### ISSUE THREAD

**Issue by slomp** [halide/Halide#2582]
In the following snippet:

```cpp
Halide::Buffer<uint8_t> onion = Halide::Runtime::Buffer<uint8_t>::make_scalar();
onion() = 42;

const Halide::Buffer<> garlic = onion;

Var x, y, k;
Func pepper3d;
pepper3d(x, y, k) = garlic();
```

`garlic()` appears to be resolved as an "lhs" expression and fails with the following:
```
error C2338: Cannot use operator() on Buffer<void> types
```

The workaround is to call it like this:
```cpp
pepper3d(x, y, k) = garlic( std::vector<Expr>{} );
```

Known to happen on VisualStudio 2015 and 2017.

**Comment by steven-johnson** [comment]
I don't think this is a bug: for Buffer<void> we don't know the type at compile time, so we can't make operator() return something predictable. What type would expect to be returned?

**Comment by zvookin** [comment]
There are two flavors of operator(). The first takes ints as its arguments and does an immediate indexing operation. Due to SFINAE, this is only available for Buffer's with a concrete type. The other flavor takes an Expr and returns one as well. The second one is what is being explicitly selected in the last statement. Note that the first method compiles the constant "42" into the code where the second one compiles in the buffer. (It is possible optimization cleans up the buffer, but it is not guaranteed. The Buffer can also be passed as a n argument to realize et al at which point it would be an input parameter.)

**Comment by slomp** [comment]
Just adding to Zalman's feedback:
Note that `pepper3d` is a `Func`, with `garlic()` being used as a "rhs" expression.
While Buffer<void> does not have a type per se (can't be used as an lhs), it does store a Halide type internally which can be used to construct an `Expr` that can then be bound to a `Func`.

The following, for example, works just fine (assuming `onion` above is an 1D buffer):
```cpp
pepper3d(x, y, k) = garlic(x);
```

**Comment by steven-johnson** [comment]
Ahhh, right, I missed that this was Halide::Buffer and not Halide::Runtime::Buffer.

**Comment by zvookin** [comment]
The `onion()` call does not return an `Expr` but a `uint8_t`. It isn't clear that is what you want, but the design is to favor direct indexing if the arguments are immediate and Expr production if any of them are Exprs (or Var/RVar, etc.). With no argument, there is no way to make that decision. It may be possible to get the template system to always compile the Expr version if the Buffer doesn't have a concrete type, but I'm sort of finding the thing a bit too confusing already...

**Comment by slomp** [comment]
Yes, `onion` returns an `uint8_t` because `onion` is a `Buffer<uint8_t>`; no argument there, that's expected.

Anyway, my end goal is to have some handy code like the following (used for quick prototipation):
```cpp
template<typename T>
Halide::Func Identity(const Halide::Buffer<T>& image, const char* name=nullptr)
{
    using namespace Halide;
    Func I = (name) ? Func(name) : Func(image.name()) ;
    Var x("x"), y("y"), k("k");
    const auto dimensions = image.dimensions();
    switch (dimensions)
    {
    case 0 :  // constant
        //I(x, y, k) = image();                      // <- operator() can't be used without arguments with Buffer<void>...
        I(x, y, k) = image( std::vector<Expr>{} );   // <- must call it like this instead
        break;
    case 1 :  // 1D LUT or multi-channel constant
        I(x, y, k) = image(x);
        break;
    case 2 :  // monochromatic image or multi-channel 1D LUT
        I(x, y, k) = image(x, y);
        break;
    case 3 :  // multi-channel image
        I(x, y, k) = image(x, y, k);
        break;
    default : // unknown/invalid
        break;
    }
    return(I);
}
```
```cpp
Buffer<> image = ...;
Func filter = MakeFilter( Border(input), ... );
filter.realize(output);
```

**Comment by slomp** [comment]
You have a good point: if solving this requires fighting the type system is some non intuitive way, then the juice isn't worth the squeeze and we can archive (and document) this issue.
It is a very minor issue and not super useful anyway :)

**Comment by zvookin** [comment]
Even if we made oeprator()() work in this case, I think you want your `Identity` function to use the Expr version, not the direct index version. I would expect having it go the constant route would cause unexpected issues further down the line. So you'd have to write the code that way anyway. (The counter argument is Halide is a functional language so eval now is the same as eval later. I find this a little um.... optimistic....)

I'm not happy about it, but wiring up something that does the unexpected thing might end up costing more time via the surprise later...

**Comment by zvookin** [comment]
Also, I'm starting to think we might want different syntax for direct access vs. `Expr` access. This would involve using a method for direct access via concrete coordinates. However, I think Andrew considers ease of use for `Buffer<T>` in plain C++ programming somewhat more important than I do.

**Comment by abadams** [comment]
**Automated triage note:** this issue was flagged by Claude as likely safe to close, for the reason below. The reasoning was checked against the current source, and a second pass tried and failed to refute it, but no human verified this specific text. **If the reason is wrong, please reopen the issue and say so** — that is the expected outcome for some fraction of these, and no offence taken.

**Reason (CLOSE_WONTFIX):** The report still reproduces exactly as filed, but the thread concluded not to fix it: the zero-arg case is ambiguous between direct indexing and Expr production, zvookin argued a fix would silently do the surprising thing, and the reporter agreed the juice isn't worth the squeeze and suggested archiving.

**Evidence:** Reproduced against current main: a TU compiling `pepper3d(x,y,k) = garlic();` with `Buffer<> garlic` fails at src/runtime/HalideBuffer.h:2120 "static assertion failed: Cannot use operator() on Buffer<void> types", reached through src/Buffer.h:574-576. The `garlic(std::vector<Expr>{})` workaround (src/Buffer.h:591) still works.

Still reproduces, but per the discussion this is a deliberate ambiguity in operator(): with zero arguments there is no way to choose between direct indexing and Expr production, and the Expr form is what callers actually want. Use `buf(std::vector<Expr>{})` for the Expr form. Closing as wontfix.
